### PR TITLE
Factor althold logic into stabilization

### PR DESCRIPTION
**Pull request by mlyle**
Has tested in simulator.

Remaining work items

* [x] Cleaning up config dirtiness
* [x] Making sanity check sane
* [x] Not building it on targets that'll never need it
* ~~Moving VTOLPathFollower to use this method of altitude control~~ [defer]
* [x] Sim testing
* [x] Bench testing
* [x] Flight testing

### CODE REVIEW

**Comment by mlyle** [comment]
Tested on dtfc to not allow altitudehold module.

**Comment by mlyle** [comment]
Tested on sparky2 to apparently do the right thing with thrust when held in hand.

**Comment by mlyle** [comment]
Retested on DTFc to properly mask althold mode.

**Comment by mlyle** [comment]
Tested on Revolution with no baro to properly mask althold mode.

**Comment by mlyle** [comment]
Confirmed on seppuku w/ baro hacked out that the mode is sanitychecked masked...

**Comment by mlyle** [comment]
Need to confirm on bench that F3 CPU use is OK in althold mode, since this does more work.  Need to flight test.

**Comment by mlyle** [comment]
SPRF3E uses 68% CPU in leveling, 73% CPU in althold, so not too bad.  (definitely worse than before, but hey, sampling faster is good right?)

Confirm SPRF3E with baro works on bench, without baro configerrors.

File: flight/Modules/Stabilization/stabilization.c [hunk 1]
```diff
@@ -375,6 +392,190 @@ static void calculate_attitude_errors(uint8_t *axis_mode, float *raw_input,
...
+		last_nonzero_thrust_time = 0;
+
+		return 0.0f;
+	}
+
+#ifdef TARGET_MAY_HAVE_BARO
+	bool do_alt_control = false;
+	bool do_vs_control = false;
+
+	static bool old_vs_control;
```
Comment (tracernz) [inline flight/Modules/Stabilization/stabilization.c]: This confused me, this is just the previous state, not an old algorithm or something? Suggest prev_vs_control, as per other previous state vars.

File: flight/Modules/Stabilization/stabilization.c [hunk 2]
```diff
@@ -375,6 +392,190 @@ static void calculate_attitude_errors(uint8_t *axis_mode, float *raw_input,
...
+	} else if (flightStatus->Armed != FLIGHTSTATUS_ARMED_ARMED) {
+		last_thrust_pos = true;
+		last_nonzero_thrust_time = 0;
+
+		return 0.0f;
+	}
+
+#ifdef TARGET_MAY_HAVE_BARO
+	bool do_alt_control = false;
+	bool do_vs_control = false;
```
Comment (tracernz) [inline flight/Modules/Stabilization/stabilization.c]: vertspeed_control, alti_control?

File: flight/Modules/Stabilization/stabilization.c [hunk 3]
```diff
@@ -375,6 +392,190 @@ static void calculate_attitude_errors(uint8_t *axis_mode, float *raw_input,
...
+
+			// Add ability to scale up the amount of compensation to achieve
+			// level forward flight
+			fraction = powf(fraction, (float) altitudeHoldSettings.AttitudeComp / 100.0f);
+
+			// Dividing by the fraction remaining in the vertical
+			// projection will attempt to compensate for tilt.
+			// If the fraction is starting to go negative we are
+			// inverted and should shut off throttle (but keep
+			// stabilization).
```
Comment (tracernz) [inline flight/Modules/Stabilization/stabilization.c:530]: This doesn't seem right for vehicles with symmetrical thrust?

File: flight/Modules/Stabilization/stabilization.c [hunk 4]
```diff
@@ -375,6 +392,190 @@ static void calculate_attitude_errors(uint8_t *axis_mode, float *raw_input,
...
+				0	// don't use slope-antiwindup
+				);
+
+		AltitudeHoldStateData altitudeHoldState;
+		altitudeHoldState.VelocityDesired = vs_desired;
+		altitudeHoldState.Integral = vertspeed_pid.iAccumulator;
+		altitudeHoldState.AngleGain = 1.0f;
+
+		if (altitudeHoldSettings.AttitudeComp > 0) {
+			// Thrust desired is at this point the mount desired in the up direction, we can
```
Comment (tracernz) [inline flight/Modules/Stabilization/stabilization.c]: typo in amount :P

File: flight/Modules/Stabilization/stabilization.c [hunk 5]
```diff
@@ -375,6 +392,190 @@ static void calculate_attitude_errors(uint8_t *axis_mode, float *raw_input,
...
+			// Project a unit vector pointing up into the body frame and
+			// get the z component
+			float fraction = attitudeActual.q1 * attitudeActual.q1 -
+				attitudeActual.q2 * attitudeActual.q2 -
+				attitudeActual.q3 * attitudeActual.q3 +
+				attitudeActual.q4 * attitudeActual.q4;
+
+			// Add ability to scale up the amount of compensation to achieve
+			// level forward flight
+			fraction = powf(fraction, (float) altitudeHoldSettings.AttitudeComp / 100.0f);
```
Comment (tracernz) [inline flight/Modules/Stabilization/stabilization.c]: space after cast

File: flight/Modules/Stabilization/stabilization.c [hunk 6]
```diff
@@ -390,16 +591,29 @@ static void stabilizationTask(void* parameters)
 	FlightStatusData flightStatus;
 	SystemSettingsAirframeTypeOptions airframe_type;
 
+	volatile bool settings_flag = true;
+	volatile bool flightstatus_flag = true;
```
Comment (tracernz) [inline flight/Modules/Stabilization/stabilization.c]: Suggest, _updated, flag is very ambiguous.

File: flight/Modules/Stabilization/stabilization.c [hunk 7]
```diff
@@ -390,16 +591,29 @@ static void stabilizationTask(void* parameters)
...
 	SubTrimSettingsConnectCallbackCtx(UAVObjCbSetFlag, &settings_flag);
+#ifdef TARGET_MAY_HAVE_BARO
+	AltitudeHoldSettingsConnectCallbackCtx(UAVObjCbSetFlag, &settings_flag);
+#else
+	/*
+	 * Sanity check uses this to know whether althold mode is permitted,
+	 * so we'd better not have a baro if the target doesn't "support"
+	 * one.
+	 */
+	PIOS_Assert(!PIOS_SENSORS_IsRegistered(PIOS_SENSOR_BARO));
```
Comment (tracernz) [inline flight/Modules/Stabilization/stabilization.c:619]: Would be better to put this into the sanity check IMO.

File: flight/Modules/Stabilization/stabilization.c [hunk 8]
```diff
@@ -632,7 +810,7 @@ static void stabilizationTask(void* parameters)
 		uint16_t max_safe_rate = PIOS_SENSORS_GetMaxGyro() * 0.9f;
 
 		//Run the selected stabilization algorithm on each axis:
-		for(uint8_t i=0; i< MAX_AXES; i++)
+		for (uint8_t i = 0; i < MAX_AXES; i++)
 		{
```
Comment (tracernz) [inline flight/Modules/Stabilization/stabilization.c]: If you're cleaning it up, move the curly-bracket up where it belongs. :D

**Review by tracernz — COMMENTED** [review]
Good work.

File: flight/Modules/Stabilization/stabilization.c (hunk 7 above)
Comment (mlyle) [reply flight/Modules/Stabilization/stabilization.c:619]: The sanity check is in a library and is built for all targets.  It doesn't know if MAY_HAVE_BARO is true, so there's no choice but to do it here.

File: flight/Modules/Stabilization/stabilization.c (hunk 6 above)
Comment (mlyle) [reply flight/Modules/Stabilization/stabilization.c]: OK i just moved these didn't I? :P  But fine :P

File: flight/Modules/Stabilization/stabilization.c (hunk 4 above)
Comment (mlyle) [reply flight/Modules/Stabilization/stabilization.c]: :)

File: flight/Modules/Stabilization/stabilization.c (hunk 3 above)
Comment (mlyle) [reply flight/Modules/Stabilization/stabilization.c:530]: It's from the existing althold module.  If we want to make althold work inverted, or make it make a pulse-of-negative-thrust when inverted until crossing level, we can think about it.  (Right now, with it being a unistable attitude thing, doesn't make much sense)

File: flight/Modules/Stabilization/stabilization.c (hunk 1 above)
Comment (mlyle) [reply flight/Modules/Stabilization/stabilization.c]: OK

File: flight/Modules/Stabilization/stabilization.c (hunk 2 above)
Comment (mlyle) [reply flight/Modules/Stabilization/stabilization.c]: OK

File: flight/Modules/Stabilization/stabilization.c (hunk 8 above)
Comment (mlyle) [reply flight/Modules/Stabilization/stabilization.c]: :D

File: flight/Modules/Stabilization/stabilization.c (hunk 3 above)
Comment (tracernz) [reply flight/Modules/Stabilization/stabilization.c:530]: Yup. Would be super awesome to get that working one day. 😃

File: flight/Modules/Stabilization/stabilization.c (hunk 7 above)
Comment (tracernz) [reply flight/Modules/Stabilization/stabilization.c:619]: Makes sense. :+1:

**Comment by mlyle** [comment]
Flown on Sparky2 successfully with default coefficients.  It's clearly superior to old althold in accuracy and smoothness (I guess doing it faster than 50Hz makes a difference).

Also flew conservative "vertical autotune" coeffs, which are awesome.

**Comment by mlyle** [comment]
https://gist.github.com/mlyle/5dbe97aa964d3709db3a11c189471d70

**Comment by mlyle** [comment]
Removed an inadvertent change disabling baro on seppuku (used during sanity test).

**Review by tracernz — APPROVED** [review]
Changes look good, thanks.

**Pull request merged**